### PR TITLE
Rewrote introspectMachine to be more readable

### PR DESCRIPTION
**Pull request by mattpocock**

### CODE REVIEW

**Comment by mattpocock** [comment]
@mwarger Could you check this out? Hoping this clears up some of the nasty code in there

File: packages/xstate-compiled/src/introspectMachine.ts [hunk 1]
```diff
@@ -62,19 +62,99 @@ const makeSubStateFromNode = (
...
 };
 
+class ItemMap {
+  /**
+   * The internal map that we use to keep track
+   * of all of the items
+   */
+  private map: {
+    [name: string]: { events: Set<string>; states: Set<XState.StateValue> };
+  } = {};
```
Comment (mattpocock) [inline packages/xstate-compiled/src/introspectMachine.ts:72]:
```suggestion
  } = {};

```

File: packages/xstate-compiled/src/introspectMachine.ts [hunk 2]
```diff
@@ -62,19 +62,100 @@ const makeSubStateFromNode = (
...
+
+  /**
+   * Check if one of these items is optional -
+   * passed in from above via a prop
+   */
+  private checkIfOptional: (name: string) => boolean;
```
Comment (mwarger) [inline packages/xstate-compiled/src/introspectMachine.ts:78]: I think TS 4.0 would allow this to be inferred via the constructor.

**Review by mwarger — APPROVED** [review]
LGTM - this is a great step towards standardizing the operations.

**Pull request merged**